### PR TITLE
modifing the install structure for discovery and install function

### PR DESCRIPTION
**Pull request by Rocketct**

### CODE REVIEW

File: daemon/client/client.go [hunk 1]
```diff
@@ -313,17 +313,8 @@ func main() {
...
-		Instance: instance,
-		Name:     "WiFi101",
-		Version:  "0.15.3",
-	})
-	if err != nil {
-		fmt.Printf("Error Upgrade platform: %s\n", err)
-		os.Exit(1)
-	}
+	// LIB INSTALL
+	fmt.Println("=== calling libraryInstal")
```
Comment (per1234) [inline daemon/client/client.go]:
```suggestion
	fmt.Println("=== calling LibraryInstall(WiFi101@0.15.3)")
```

File: daemon/client/client.go [hunk 2]
```diff
@@ -346,6 +337,20 @@ func main() {
 			fmt.Printf(">> INSTALL: %s\n", installResp.GetProgress())
 		}
 	}
+	// LIB UNINSTALL
+	fmt.Println("=== calling libraryUninstal")
```
Comment (per1234) [inline daemon/client/client.go]:
```suggestion
	fmt.Println("=== calling LibraryUninstall(WiFi101@0.15.3)")
```

**Review by per1234 — CHANGES_REQUESTED** [review]

File: daemon/client/client.go [hunk 3]
```diff
@@ -381,6 +382,20 @@ func main() {
...
+		Instance: instance,
+		Names:    true,
+		Query:    "audio",
+	})
+	if err != nil {
+		formatter.PrintError(err, "Error saerching for Library")
+		os.Exit(1)
+	}
+
+	fmt.Println(outputsrc)
```
Comment (cmaglie) [inline daemon/client/client.go:397]: Rename `outputsrc` to `searchResp`
Print in this way to match the rest of client.go:

```
fmt.Printf("---> %+v\n", searchResp)
fmt.Println()
```

File: cli/lib/search.go [hunk 4]
```diff
@@ -47,30 +52,58 @@ var searchFlags struct {
...
+			fmt.Sprintln("  Types: ", strings.Join(lsr.GetLatest().Types, ", ")) +
+			fmt.Sprintln("  Versions: ", strings.Replace(fmt.Sprint(Versions(lsr.GetLatest(), lsr)), " ", ", ", -1))
+	}
+	return strings.TrimSpace(ret)
+}
+
+func Versions(l *rpc.LibraryRelease, library *rpc.SearchLibraryOutput) []*semver.Version {
+	res := []*semver.Version{}
+	for _, rel := range library.Releases {
+		if rel.GetAuthor() == l.GetAuthor() {
```
Comment (cmaglie) [inline cli/lib/search.go]: The comparison should be on the Name not the Author

File: cli/lib/search.go [hunk 5]
```diff
@@ -47,30 +52,58 @@ var searchFlags struct {
...
 func runSearchCommand(cmd *cobra.Command, args []string) {
+	instance := cli.CreateInstance()
 	logrus.Info("Executing `arduino lib search`")
-	query := strings.ToLower(strings.Join(args, " "))
-
-	lm := cli.InitLibraryManager(cli.Config)
-
-	res := output.LibSearchResults{
-		Libraries: []*librariesindex.Library{},
+	arguments := strings.ToLower(strings.Join(args, " "))
```
Comment (cmaglie) [inline cli/lib/search.go]: ToLower should be done inside LibrarySearch(..)

File: cli/lib/search.go [hunk 6]
```diff
@@ -47,30 +52,58 @@ var searchFlags struct {
...
+	outputsrc, err := lib.LibrarySearch(context.Background(), &rpc.LibrarySearchReq{
```
Comment (cmaglie) [inline cli/lib/search.go]: rename `outputsrc` -> `searchResp`

File: cli/lib/search.go [hunk 7]
```diff
@@ -47,30 +52,58 @@ var searchFlags struct {
...
+		Query:    arguments,
+	})
+	if err != nil {
+		formatter.PrintError(err, "Error saerching for Library")
+		os.Exit(cli.ErrGeneric)
 	}
-	for _, lib := range lm.Index.Libraries {
-		if strings.Contains(strings.ToLower(lib.Name), query) {
-			res.Libraries = append(res.Libraries, lib)
+	if cli.OutputJSONOrElse(outputsrc.GetSearchOutput()) {
```
Comment (cmaglie) [inline cli/lib/search.go]: You're using `outputsrc.GetSearchOutput()` plenty of times, just assign it to a variable like:

```go
results := outputsrc.GetSearchOutput()
```

and use `results` instead

File: commands/lib/search.go [hunk 8]
```diff
@@ -18,59 +18,77 @@
...
+	res := []*rpc.SearchLibraryOutput{}
 
-	res := output.LibSearchResults{
-		Libraries: []*librariesindex.Library{},
-	}
 	for _, lib := range lm.Index.Libraries {
-		if strings.Contains(strings.ToLower(lib.Name), query) {
-			res.Libraries = append(res.Libraries, lib)
+		if strings.Contains(strings.ToLower(lib.Name), req.GetQuery()) {
+			release := map[string]*rpc.LibraryRelease{}
```
Comment (cmaglie) [inline commands/lib/search.go]: rename `release` -> `releases`

File: commands/lib/search.go [hunk 9]
```diff
@@ -18,59 +18,77 @@
...
+			restmp = append(restmp, searchedlib)
 		}
+		res = restmp
 	} else {
-		if len(res.Libraries) == 0 {
-			formatter.Print(fmt.Sprintf("No library found matching `%s` search query", query))
-		} else {
-			formatter.Print(res)
+		if len(res) == 0 {
+			return nil, fmt.Errorf("%s library not found", req.GetQuery())
```
Comment (cmaglie) [inline commands/lib/search.go]: I think we should not error but just return an empty list.

File: rpc/lib.proto [hunk 10]
```diff
@@ -61,4 +61,40 @@ message LibraryUpgradeAllReq {
...
+message LibrarySearchResp {
+    repeated SearchLibraryOutput search_output = 1;
+}
+
+message SearchLibraryOutput {
+    string Name = 1;
+    map<string,LibraryRelease> Releases = 2;
+    LibraryRelease latest = 3;
+}
+message LibraryRelease {
```
Comment (cmaglie) [inline rpc/lib.proto:82]: Missing a blank line between `message`

**Review by cmaglie — CHANGES_REQUESTED** [review]

File: rpc/lib.proto (hunk 10 above)
Comment (Rocketct) [reply rpc/lib.proto:82]: done

File: commands/lib/search.go (hunk 8 above)
Comment (Rocketct) [reply commands/lib/search.go]: done

File: cli/lib/search.go (hunk 7 above)
Comment (Rocketct) [reply cli/lib/search.go]: done

File: cli/lib/search.go (hunk 6 above)
Comment (Rocketct) [reply cli/lib/search.go]: done

File: cli/lib/search.go (hunk 4 above)
Comment (Rocketct) [reply cli/lib/search.go]: removed useless in this point

File: commands/lib/search.go (hunk 9 above)
Comment (Rocketct) [reply commands/lib/search.go]: i'm agree and done

File: cli/lib/search.go (hunk 5 above)
Comment (Rocketct) [reply cli/lib/search.go]: done

**Pull request merged**